### PR TITLE
Fix EZP-24586: DateAndTime field type setting 'dateInterval' can be u…

### PR DESCRIPTION
**Pull request by glye**
…sed only when setting 'defaultType' is set to 'DEFAULT_CURRENT_DATE_ADJUSTED'

https://jira.ez.no/browse/EZP-24586
Manual tests.

This fixes the bug. It also improves the error message (for when it is supposed to be shown) so that it shows the translated string the user can see on the page, rather than the internal string (DEFAULT_CURRENT_DATE_ADJUSTED).

### CODE REVIEW

**Comment by glye** [comment]
Travis is down.

**Comment by glye** [comment]
Rebased on master, now with 100% more PSR-2.

File: eZ/Publish/Core/FieldType/DateAndTime/Type.php [hunk 1]
```diff
@@ -246,19 +258,23 @@ public function validateFieldSettings($fieldSettings)
...
+                                    $fieldSettings['defaultType'] !== self::DEFAULT_CURRENT_DATE_ADJUSTED &&
+                                    $value->format('%y%m%d%h%i%s') !== '000000') {
+                                    $validationErrors[] = new ValidationError(
+                                        "Setting '%setting%' can be used only when setting '%defaultType%' is set to '%DEFAULT_CURRENT_DATE_ADJUSTED%'",
+                                        null,
+                                        array(
+                                            'setting' => $this->translator->trans('field_definition.ezdatetime.date_interval', [], 'ezrepoforms_content_type'),
+                                            'defaultType' => $this->translator->trans('field_definition.ezdatetime.default_type', [], 'ezrepoforms_content_type'),
+                                            'DEFAULT_CURRENT_DATE_ADJUSTED' => $this->translator->trans('field_definition.ezdatetime.default_type_adjusted', [], 'ezrepoforms_content_type'),
+                                        ),
```
Comment (lolautruche) [inline eZ/Publish/Core/FieldType/DateAndTime/Type.php]: You should not use Symfony services in API related stuff. `ValidationError` is already translatable. Here you just need to assign the values. Why do you need translation for this?

File: eZ/Publish/Core/FieldType/DateAndTime/Type.php [hunk 2]
```diff
@@ -246,19 +258,23 @@ public function validateFieldSettings($fieldSettings)
...
-                                    ),
-                                    "[$name]"
-                                );
-                            } elseif (!($value instanceof \DateInterval)) {
+                            if ($value instanceof \DateInterval) {
+                                // String conversion of $value, because DateInterval objects cannot be compared directly
+                                if (isset($fieldSettings['defaultType']) &&
+                                    $fieldSettings['defaultType'] !== self::DEFAULT_CURRENT_DATE_ADJUSTED &&
+                                    $value->format('%y%m%d%h%i%s') !== '000000') {
+                                    $validationErrors[] = new ValidationError(
```
Comment (lolautruche) [inline eZ/Publish/Core/FieldType/DateAndTime/Type.php]: Wrong format :-). Should be:

```php
if(
    isset($fieldSettings['defaultType'])
    && $fieldSettings['defaultType'] !== self::DEFAULT_CURRENT_DATE_ADJUSTED
    && $value->format('%y%m%d%h%i%s') !== '000000')
) {
    // ...
}
```

File: eZ/Publish/Core/FieldType/DateAndTime/Type.php [hunk 3]
```diff
@@ -246,19 +258,23 @@ public function validateFieldSettings($fieldSettings)
...
-                                    null,
-                                    array(
-                                        'setting' => $name,
-                                        'defaultType' => 'defaultType',
-                                        'DEFAULT_CURRENT_DATE_ADJUSTED' => 'DEFAULT_CURRENT_DATE_ADJUSTED',
-                                    ),
-                                    "[$name]"
-                                );
-                            } elseif (!($value instanceof \DateInterval)) {
+                            if ($value instanceof \DateInterval) {
```
Comment (lolautruche) [inline eZ/Publish/Core/FieldType/DateAndTime/Type.php]: `DateInterval` is already imported.

Comment (glye) [inline eZ/Publish/Core/FieldType/DateAndTime/Type.php]: @lolautruche meaning...?

**Comment by glye** [comment]
Fixed CS errors, use short array syntax all over, namespace fix for DateTime and DateInterval.

**Comment by lolautruche** [comment]
+1 if Travis agrees :)

**Comment by andrerom** [comment]
+1, but would be better if the PR did not contain coding standard changes as well.

**Comment by glye** [comment]
It won't happen again :)

**Pull request merged**